optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,156 @@
+        min_early_stopping_rate_high:
+            The end point of the minimum early stopping rate for ``SuccessiveHalvingPruner``.
+    """
+
+    def __init__(
+            self,
+            min_resource=1,
+            reduction_factor=3,
+            min_early_stopping_rate_low=0,
+            min_early_stopping_rate_high=4
+    ):
+        # type: (int, int, int, int) -> None
+        self._pruners = []  # type: List[SuccessiveHalvingPruner]
+        self._reduction_factor = reduction_factor
+        self._resource_budget = 0
+        n_pruners = min_early_stopping_rate_high - min_early_stopping_rate_low + 1
+        self._n_pruners = n_pruners
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hyperband has {} brackets'.format(self.n_pruners))
+
+        for i in range(n_pruners):
+            bracket_resource_budget = self._calc_bracket_resource_budget(i, n_pruners)
+            self._resource_budget += bracket_resource_budget
+            self._bracket_resource_budgets.append(bracket_resource_budget)
+
+            # N.B. (crcrpar): `min_early_stopping_rate` has the information of `bracket_index`.
+            min_early_stopping_rate = min_early_stopping_rate_low + i
+
+            _logger.debug(
+                '{}th bracket has minimum early stopping rate of {}'.format(
+                    i, min_early_stopping_rate))
+
+            pruner = SuccessiveHalvingPruner(
+                min_resource=min_resource,
+                reduction_factor=reduction_factor,
+                min_early_stopping_rate=min_early_stopping_rate,
+            )
+            self._pruners.append(pruner)
+
+    def prune(self, study, trial):
+        # type: (Study, FrozenTrial) -> bool
+
+        i = self.get_bracket_id(study.study_name, trial.number)
+        _logger.debug('{}th bracket is selected'.format(i))
+        bracket_study = _BracketStudy(study, i)
+        return self._pruners[i].prune(bracket_study, trial)
+
+    def _calc_bracket_resource_budget(self, pruner_index, n_pruners):
+        # type: (int, int) -> int
+        n = self._reduction_factor ** (n_pruners - 1)
+        budget = n
+        for i in range(pruner_index, n_pruners - 1):
+            budget += n / 2
+        return budget
+
+    def get_bracket_id(self, study_name, trial_number):
+        # type: (str, int) -> int
+        """Computes the index of bracket for a trial of ``trial_number``.
+
+        The index of a bracket is noted as :math:`s` in
+        `Hyperband paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_.
+        """
+
+        n = hash('{}_{}'.format(study_name, trial_number)) % self._resource_budget
+        for i in range(self.n_pruners):
+            n -= self._bracket_resource_budgets[i]
+            if n < 0:
+                return i
+
+        raise RuntimeError
+
+    @property
+    def resource_budget(self):
```

Comment:
Is any logic expected to be added in the near future? Looking at the current implementation, I believe it is slightly overengineered(non-Python) to have a custom getter instead of a plain (public) attribute.

Reply:
> Is any logic expected to be added in the near future?

No. I will remove this property and the below one as I don't have any usage.